### ISSUE TITLE
Override Stream.CopyTo on DeflateStream

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#29478]
Stream.CopyTo was made virtual in a previous release of .NET Core.  We override CopyToAsync on DeflateStream, but we don't currently override CopyTo.

**Comment by tdinucci** [comment]
I'd be happy to take this along with #29479

**Comment by stephentoub** [comment]
Thanks.

> What would you prefer I do?

I'll still take a look... just haven't gotten to it yet.

> One concern I have about this though is how this will play out when desktop UI support comes with .Net Core v3. I'm assuming these are still going to be STA?

Remember, without this change, when FileStream is in this async mode, CopyTo is already doing sync-over-async, in that the base Stream.CopyTo is just going to call FileStream.Read repeatedly, and FileStream.Read will end up just doing FileStream.ReadAsync().GetAwaiter().GetResult().  That's been the case since the beginning-ish of .NET.  This change just makes that blocking more efficient by doing it once per copy rather than N times per copy.

**Comment by tdinucci** [comment]
This one's for DeflateStream :) So this would be introducing a risk that isn't currently present.

**Comment by stephentoub** [comment]
Yeah, I added an EDIT almost immediately after writing my previous comment, once I saw this was for DeflateStream.

> So this would be introducing a risk that isn't currently present.

This one wouldn't be based on CopyToAsync; rather, it would follow your second bullet: "Replicating the way CopyToAsync is implemented but altered to run completely synchronously."

The primary benefits of the CopyToAsync implementation are called out in the PR notes in https://github.com/dotnet/corefx/pull/13184.  Some of those still apply in CopyTo.  Additionally, I expect one of the biggest boons would just be in respecting the supplied bufferSize argument.  Today there's no way to tell DeflateStream how big an internal buffer to create, and while I expect that might show up in the future, that's already dialable today with CopyToAsync and the same could be done with CopyTo.  The size of that buffer can make a significant difference on compression throughput.  You might try your experiments with a really large buffer, e.g. compressing/decompressing a 2GB file with a 512K bufferSize argument, before and after your change (maybe first verifying the expected impact with CopyToAsync).
